### PR TITLE
Tetiana's blogpost outline

### PR DESCRIPTION
**Pull request by tetianadadakova**
Initial example for my blogpost

### CODE REVIEW

**Comment by abbywlsn** [comment]
Overall, this is a great project! I am really impressed by what you've accomplished. Here are my suggestions moving forward: 

There are a couple of spelling and grammar mistakes (no worries we are all coders, not English majors). I know VS code has a spellchecker and most editors also do. I'd look into one of those. If you're writing your blog post in the GitHub editor, Grammarly is also great. 

Here are a couple of fixes for some of the things I found though. 

I would capitalize your title to be Genome Change Logging Using Range Map 

In your first code block, you have atarting in your last comment, I think you mean starting. 
In another code block you also use {key : {val, inseet}}, this should be {key : {val, insert}} (probably)

You also say "the advantages of such approach". I would write "the advantages of this approach" 

"Inserted into the map dusing" should be during 

There's a couple more but I think that just proofreading and a spellchecker will fix most of it! 

Now for the actual material: 

I think including some diagrams or even a gif would really help explain what genome insertions, deletions, etc are. Here is a link to one I found, though I'm sure there are better ones. 

https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.docsity.com%2Fen%2Fnews%2Feducation-2%2Finteresting-gifs-related-molecular-biology-genetics%2F&psig=AOvVaw3jHTs06lqOhhx1rgUyGUxf&ust=1596218564757000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCPDsj9fH9eoCFQAAAAAdAAAAABAD

In terms of explanation, I would explain what a naive implementation is. I would also mention that copying the entire genome, especially ones that are really long, is a time-consuming process. 

I would also just add a reminder (for non-code people) that you index from zero. 

I also think that adding the compile time/ run time comparison at the end of the post will be an interesting piece of info. 

Overall, just adding a little more explanation of concepts will really round out the post. 

This is a great project though and you've done really impressive work! I'm excited to see the final product! Nice job!

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+---
+layout: post
+title: "Future blogpost about MABE"
+date: 2020-07-15
+author: Tetiana
+---
+
+# Genome change logging using range map  
+
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:9]: I would include a brief intro/overview as to what your project was for this summer (and any goals you had) before diving straight into the details

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+For more information about MABE (Modular Agent-Based Evolution platform) see [this introduction](https://szendejo.github.io/waves/blog/Team-MABE.html).
+
+## Genome Class
+
+### Naïve Implementation
+
+The naïve implementation of the genome is using a vector data structure from the standard library. The overwrite, insert and remove operations are, therefore, implemented using the standard library algorithms, specifically:
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:16]: I would specifically mention that using a vector data structure is a way of representing a genome. So maybe something like: "The naïve way to represent a genome with code is ..." 

I would also include a brief introductory sentence as to what the overall goal of the genome class is

Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:16]: I would change the second line to "In this naïve implementation, the overwrite, insert and remove operations are thus implemented using the following standard library algorithms:" just to make it a bit more clear

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+The advantages of such approach include:
+* All the sites are in contiguous memory -> fast operations due to cache-friendliness (e.g. random access or iterations)  
+* Use of C++ standard library data structures -> code is simple, readable, expressive and optimized for performance  
+  
+However, there are also disadvantages:  
+* Every generation, the whole genome (the whole `sites` vector) is copied and then the mutations are applied to it -> In a common situation of large genome and low mutation rates, it means copying a lot of values that didn't change  
+* The `insert()` and `erase()` algorithms have linear time complexity -> inefficient time  
+  
+  **The goal of this project was to investigate if storing only the mutations (as opposed to storing the whole genome) would provide a better time and memory performance.**
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:47]: I might move this (or a version of this) to the top of the section to give it a clearer introduction

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+### Optimized Implementation Using Change Log  
+
+One of the ways to improve the time complexity as well as optimize for memory use is to have a change log. The change log will keep track of the mutations that occurred between the parent and the offsprings over generations. This means only storing the differences between parent genome and it's offsprings as opposed to storing the whole genome for every offspring.
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:51]: A small thing, but I think the proper markdown style is to have each sentence be on its own line for readability purposes

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+One of the ways to improve the time complexity as well as optimize for memory use is to have a change log. The change log will keep track of the mutations that occurred between the parent and the offsprings over generations. This means only storing the differences between parent genome and it's offsprings as opposed to storing the whole genome for every offspring. 
+
+As we've seen above, the algorithm has to support the following mutations:
+* Overwrite - the values at one or more sites is overwritten by a new value
+* Insert - one or more sites are inserted into at specific location
+* Remove - one or more sites are removed at specific location
+
+My implementation consists of two maps from the standard library:
+* **change_log** is implemeted as std::map and contains the information about the **number of inserted and removed sites**. It is used to calculate the relationship between a particular site in the offspring genome and either the parent genome or the newly inserted values stored in the segments_log (see next)
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:59]: I would surround std::map with tick marks to make it appear as code for better readability (`std::map`)

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+* **segments_log** is implemented as std::unordered_map and stores the segments that were inserted into the map dusing mutation
+
+Each genome will have it's own change_log and segments_log, which in combination with the parent genome will allow the random access to any value in the offspring genome as well as the reconstruction of complete offsping genome or a part of it as a contiguous memory block of necessary sites.
+
+One important detail of the change_log is that it doesn't store every every deleted or inserted index. Instead, to optimize for memory use, it stores only one index for each range of a particular shift in indices due to insertion of deletion (see example below). I.e. each key in the cahnge_log represents all the keys in the range from the current key until the nex key.
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:64]: typo: cahnge_log should be change_log

Comment (Oliver-BE) [reply blog/_posts/2020-07-15-Tetiana-Dadakova.md:64]: also: nex should be next

File: blog/_posts/2020-07-15-Tetiana-Dadakova.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+E.g. updating our previous map with an insertion of 3 elements {20, 21, 22} at index 7:
+```
+{3 : {-2, false}}
+{5 : {-5, false}
+{7 : {0, true}}
+```
+
+Now we see that starting from index 7 there are no values in the parent genome, we need to store inserted values somewhere else. In addition to range map, another data structure is needed to store the inserted segments. I decided to use the std::unordered_map, as it allows constant time access by key. Let's see an example of how change_log and segments_log will work together to store mutations.
```
Comment (Oliver-BE) [inline blog/_posts/2020-07-15-Tetiana-Dadakova.md:149]: I would again make std::unordered_map into `std::unordered_map`

**Comment by Oliver-BE** [comment]
Nice work! This is very well organized/layout and has tons of useful examples (and well-commented code) that clearly explain the work you did over the summer. The biggest suggestion I have is to flesh out your introduction a little more so that readers are oriented as to the project you worked on and what the goals of that project were. I only had very minimal suggestions for things to add/things to format. 

I also think the gifs/visuals that you mention adding will be a great addition and make your examples more understandable. I just downloaded the "Spell Right" VS Code extension (per Abby's suggestion) and it helped me catch a bunch of small typos that I hadn't noticed.

**Pull request merged**